### PR TITLE
kube rebase

### PR DESCRIPTION
**Pull request by deads2k**
Code compiles and the master can start, but there are severe problems because of behavioral changes with encoding/decoding, RESTMappings for types, API Group registration, and API Server implementation changes.  In addition, we're pulling in a new etcd client which is incompatible with our old one, so that will need some work too.

I've called out interesting and dangerous pieces.

This is currently broken because `runtime.RawExtension` no longer gets converted and encoded during encoding.  Well, it looks like templates are now hosed because of this problem.

- [x] compile
- [x] test-go
- [x] test-integration
- [x] test-cmd
- [x] test-end-to-end

### CODE REVIEW

**Comment by deads2k** [comment]
@soltysh @liggitt @pweil- if you want to start looking.

**Comment by deads2k** [comment]
@smarterclayton please vet `IMPORTANT: runtime.Object conversion method that works-ish for now`, particularly the comment.  I'm pretty sure this is solvable in a predictable and controllable way that would make these changes less painful.

**Comment by deads2k** [comment]
@bparees @csrwng take a look at `IMPORTANT: template encoding no longer applies conversions on contents`.  Template conversion and encoding no longer ripples down into the contained objects.  This means that everywhere that is building a template in our codebase needs to be updated.  I've added a helper to perform the "default" sort of conversion that we're doing now.  

I'll update the scheme to build a full list of all available `GroupVersion`s to make this a little easier as I go on, but this is a significant change.

**Comment by soltysh** [comment]
You can safely remove all `Copyright 2015 The Kubernetes` from all the places we have this, a few introduced in your rebase, rest are old timers.

**Comment by deads2k** [comment]
@kargakis see `I hate generators so much`.  They should die.

**Comment by soltysh** [comment]
There's one slight problem with k8s bump here, it's using 9da202e242d8ceedb549332fb31bf1a933a6c6b6 which is the id of last UPSTREAM commit, but not the actual k8s we're basing off. I'd rather this point to actual base k8s commit, imho.

**Comment by deads2k** [comment]
@smarterclayton @miminar Can you help with `hack/test-integration.sh TestV2RegistryGetTags`?  I don't know enough about what's happen in that test to quickly diagnose it.

**Comment by deads2k** [comment]
rebased for easier review.
 1. boring will be squashed down, but its separated by category if you want to look through it
 1. important should be reviewed.  These are important to have as history for blame. Behavior changes in these.
 1. interesting should be reviewed.  These are worth looking at, but I'll squash them all in before I merge.
 1. DANGER.  This scared me.  I've left a lot of cruft in place around them in case I have to revert.

**Comment by deads2k** [comment]
@soltysh I did not make much progress on the unit tests, but I did successfully make the fuzzer pass.  

Need to make sure that @liggitt is ok with "INTERESTING: eliminate v1beta3 round trip", but I'm not seeing what it buys us except more work.  We need to make sure that v1beta3 can come into internal, but we don't care about the other direction.  Fuzzing doesn't help, because our problem isn't transfering fields, its having a valid object afterwards.

**Comment by kargakis** [comment]
> Need to make sure that @liggitt is ok with "INTERESTING: eliminate v1beta3 round trip", but I'm not seeing what it buys us except more work. We need to make sure that v1beta3 can come into internal, but we don't care about the other direction. Fuzzing doesn't help, because our problem isn't transfering fields, its having a valid object afterwards.

Then why not run validation after fuzzing? We would need fuzzing to specify valid values in the first place, right?

**Comment by deads2k** [comment]
> Then why not run validation after fuzzing? We would need fuzzing to specify valid values in the first place, right?

Correct.  That's not impossible, but it is substantial work.

**Comment by deads2k** [comment]
All tests pass! Picking up all recent UPSTREAM commits, updating the fork, updating the bump, rebasing on master, and then we're good.

**Comment by soltysh** [comment]
Yay!!! Friday is perfect for reviews :-)

**Comment by deads2k** [comment]
Rebased.  I expect some tests to fail, I'll fix them in the morning.

**Comment by deads2k** [comment]
@danmcp did I anger travis?

**Comment by deads2k** [comment]
@soltysh On the name of the bump commit.  Since I've bumped off of a fork, we don't see addition upstream patches.  Using the commit of the fork is the only way to accurately know what's in the level.  If I used the kube commit, you'd have no idea which patches are included.

**Comment by deads2k** [comment]
[test]

**Comment by soltysh** [comment]
Makes sense. Stil it would be good to stress out this change
significantly,  since there will be people looking into the commit id and
either getting false info from picked commits or not found for the carries.
Just a side note.

**Comment by deads2k** [comment]
rebased again [test]

**Comment by pweil-** [comment]
ok, made it through.  This PR has 100% of your daily allowance of FaceMelt.  You've addressed all my comments.  

My only remaining concern is the suspect etcd error checking that you found in a test.   If it's not returning a 100 code for not found for some reason that's scary since we rely on that check in some places.  If it's isolated to the test case somehow then I guess it's ok and I'm at LGTM state.

@soltysh @liggitt - your turn!

**Comment by deads2k** [comment]
> My only remaining concern is the suspect etcd error checking that you found in a test. If it's not returning a 100 code for not found for some reason that's scary since we rely on that check in some places. If it's isolated to the test case somehow then I guess it's ok and I'm at LGTM state.

It was an old client thing.  I can make a <drop> patch for now and we'll remove it when we remove all traces of the old client.

@pweil- if I go green are you comfortable with me merging on your review?

**Comment by deads2k** [comment]
rebase, another upstream bug fixed, and squashed

**Comment by pweil-** [comment]
> @pweil- if I go green are you comfortable with me merging on your review?

Let's get it in and clean up whatever falls out.

Any objections @smarterclayton @liggitt ?

**Comment by deads2k** [comment]
jenkins is here: https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/872/

**Comment by liggitt** [comment]
> Let's get it in and clean up whatever falls out.
> Any objections @smarterclayton @liggitt?

As the most recent rebase is currently known as "Jordan's rebase"[1], I am in favor of letting someone else have the honor of "most recent rebase" as quickly as possible. I'll still plan to go through the changes.

[1] https://botbot.me/freenode/openshift-dev/2016-02-05/?msg=59461437&page=5

**Comment by liggitt** [comment]
pull open for `UPSTREAM: openshift/openshift-sdn: <drop>: minor updates for kube rebase`?

**Comment by deads2k** [comment]
> pull open for UPSTREAM: openshift/openshift-sdn: <drop>: minor updates for kube rebase?

Not yet.  They need to have their kube bumped, then these changes, then a rebump into openshift with the correct hash.

**Comment by deads2k** [comment]
[merge] (maybe it will get lucky)

**Comment by deads2k** [comment]
rebased again.

**Comment by deads2k** [comment]
re[test]

**Comment by liggitt** [comment]
one test failed with
```
SUCCESS after 0.407s: hack/../test/cmd/templates.sh:71: executing 'oc get pod/template-type-precision -o json' expecting success and text 'patch comment'
Running hack/../test/cmd/templates.sh:72: executing 'oc delete template/template-type-precision' expecting success...
SUCCESS after 0.471s: hack/../test/cmd/templates.sh:72: executing 'oc delete template/template-type-precision' expecting success
Running hack/../test/cmd/templates.sh:73: executing 'oc delete pod/template-type-precision' expecting success...
FAILURE after 0.463s: hack/../test/cmd/templates.sh:73: executing 'oc delete pod/template-type-precision' expecting success: the command returned the wrong error code
There was no output from the command.
Standard error from the command:
Error from server: pods "template-type-precision" not found
```

**Comment by smarterclayton** [comment]
I can't recreate it locally on my machine.

**Comment by smarterclayton** [comment]
With Go 1.5 that is

**Comment by liggitt** [comment]
it was a flake... the other test run got past that point

**Comment by smarterclayton** [comment]
Trying to recreate it locally.

**Comment by smarterclayton** [comment]
Because I don't know how in the world that can flake

**Comment by smarterclayton** [comment]
Definitely cosmic races (like cosmic rays but faster) - ran templates.sh
and unit tests in parallel for 30 minutes on both go 1.4 and go 1.5 and
didn't see it.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/apimachinery/registered/registered.go [hunk 1]
```diff
@@ -24,8 +24,10 @@ import (
 	"strings"
```
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/apimachinery/registered/registered.go]: Can I drop this in 1.4?

Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/apimachinery/registered/registered.go]: > Can I drop this in 1.4?

No.  You'll still need to have `oapi` function.  We're carrying this until those types not longer need registration.

**Pull request merged**